frontend/src/app/pages/exercises/exercise/shared/exercise-statistics/log-table/log-table.component.ts

```diff
@@ -1,11 +1,188 @@
+                            name: availableSpecifier.name,
+                            identifier: availableSpecifier.specifier,
+                            color: availableSpecifier.color,
+                            backgroundColor: availableSpecifier.backgroundColor,
+                        })),
+                ]
+            )
+        );
+    }
+
+    public get filteredLogEntries() {
+        const predicate = (logEntry: LogEntry) =>
+            this.filters.every((filter) => {
+                if (filter.specifiers.length === 0) {
+                    return logEntry.tags.some(
+                        (tag) => tag.category === filter.category
+                    );
+                }
+
+                return filter.specifiers.some((specifier) =>
+                    logEntry.tags.some(
+                        (tag) =>
+                            tag.category === filter.category &&
+                            tag.specifier === specifier.specifier
+                    )
+                );
+            });
+
+        return this.logEntries.filter(predicate);
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        if (!('logEntries' in changes)) return;
+
+        this.knownCategories = {};
+
+        // Process all tags in reverse order to use the latest available display name and color
+        [...this.logEntries].reverse().forEach((logEntry) => {
+            logEntry.tags.forEach((tag) => {
+                if (!(tag.category in this.knownCategories)) {
+                    this.knownCategories[tag.category] = {};
+                }
+
+                const knownCategory = this.knownCategories[tag.category]!;
+
+                if (!(tag.specifier in knownCategory)) {
+                    knownCategory[tag.specifier] = {
+                        specifier: tag.specifier,
+                        name: tag.name,
+                        color: tag.color,
+                        backgroundColor: tag.backgroundColor,
+                    };
+                }
+            });
+        });
+    }
+
+    addCategory(category: string) {
+        if (!this.filters.some((filter) => filter.category === category)) {
+            this.filters.push({ category, specifiers: [] });
+        }
+    }
+
+    removeCategory(category: string) {
+        const categoryIndex = this.filters.findIndex(
+            (filter) => filter.category === category
+        );
+
+        if (categoryIndex !== -1) {
+            this.filters.splice(categoryIndex, 1);
+        }
+    }
+
+    clearFilters() {
+        this.filters = [];
+    }
+
+    addSpecifierToCategory(specifier: string, category: string) {
+        const categoryFilter = this.filters.find(
+            (filter) => filter.category === category
+        );
+
+        if (!categoryFilter) return;
+
+        const specifierIndex = categoryFilter.specifiers.findIndex(
+            (filter) => filter.specifier === specifier
+        );
+
+        if (specifierIndex === -1) {
```

Comment:
This could be a `categoryFilter.specifier.every`

Reply:
I'd rather use `some` and move the negation of the condition out of the predicate. What do you think?

Reply:
Changed it in 8dcb395